### PR TITLE
add UDF nodes and button

### PR DESCRIPTION
**Pull request by AyobamBalogun**

### CODE REVIEW

File: app/index.html [hunk 1]
```diff
@@ -204,6 +206,10 @@ <h3>Please fill in connection information below and sign in for access.</h3>
                                 <li class="selected"><span class="space">Spaces</span></li>
                                 <li><span class="device">Devices</span></li>
                                 <li><span class="sensor">Sensors</span></li>
+                                <li><span class="userdefinedfunctions">UserDefinedFunctions</span></li>
```
Comment (lyrana) [inline app/index.html:212]: This text is displayed to the user? Let's match how we display in the docs: User-defined functions

File: app/scripts/spaces.js [hunk 2]
```diff
@@ -181,6 +187,13 @@ function getRequestUrlWithoDataSkip(type) {
         case "sensor":
             url = url + "sensors" + "?$top=" + oDataTop + "&$skip=" + oDataSkips[type];
             break;
+       case "userdefinedfunctions":
+            url = url + "userdefinedfunctions" + "?$top=" + oDataTop + "&$skip=" + oDataSkips[type];
+            break;
+
```
Comment (lyrana) [inline app/scripts/spaces.js]: remove extra line

File: app/scripts/spaces.js [hunk 3]
```diff
@@ -268,6 +283,13 @@ function postDigitalTwinsObject(objectType, jsonData, fComplete) {
                 case "sensor":
                     url = url + "sensors";
                     break;
+               case "userdefinedfunctions":
+                    url = url + "userdefinedfunctions";
+                    break;
+
```
Comment (lyrana) [inline app/scripts/spaces.js]: here as well, I think two more below

File: app/scripts/infopanel.js [hunk 4]
```diff
@@ -20,25 +20,44 @@ function showInfoPanel(d) {
...
   $("#panelLoaderIcon").show();
   $("#infoPanel > header")
-    .removeClass("space device sensor")
+    .removeClass("space device sensor userdefinedfunctions matchers")
     .addClass(d.type);
   $("#infoPanel h2")
-    .text(d.label)
-    .removeClass("space device sensor")
+    .text(d.label||d.name)
+      .removeClass("space device sensor userdefinedfunctions matchers")
```
Comment (lyrana) [inline app/scripts/infopanel.js]: >   [](start = 4, length = 2)

nit: check formatting, I think there're extra spaces, also down below for the breaks

File: app/style.css [hunk 5]
```diff
@@ -478,6 +487,32 @@ ul.action-menu li ul {
...
+        transform: rotate(90deg);
+        font-size: 12px;
+        margin-right: 12px;
+        display: inline-block;
+        position: relative;
+        top: -3px;
+        left: 3px;
+    }
+
+
```
Comment (lyrana) [inline app/style.css]: nit: might want to do a pass through this file as well to match existing formatting

File: app/index.html [hunk 6]
```diff
@@ -262,6 +268,10 @@ <h2 class="space">Space</h2>
                         <li id='objectAddButton' class='add'>Add Child Space</li>
                         <li id='objectAddDeviceButton' class='add'>Add Device</li>
                         <li id='objectAddSensorButton' class='addSensor'>Add Sensor</li>
+
+                        <li id='objectAddUserDefinedFunctionsButton' class='addUDF'>Add UDF</li>
```
Comment (andrew-buckley) [inline app/index.html]: Check docs, not sure abbreviations should be used.

File: app/scripts/graph.js [hunk 7]
```diff
@@ -429,15 +431,15 @@ function updateGraphVisualizer(source, animated) {
...
+            return d.label || d.name;
          })
         .attr("style", "clip-path: url(#nodeTextClipPath);")
         .attr("pointer-events", "none");
 
     // Create the drag indicator that shows when item can be dragged there..
     // We cannot drag sensors.. so we don't create one if the node is a sensor.
     var nodeDragIndicator = nodeEnter.filter(function(d){
-            return (d.type != "sensor");
+        return (d.type != "sensor");
```
Comment (andrew-buckley) [inline app/scripts/graph.js:442]: Check spacing here.

File: app/scripts/infopanel.js [hunk 8]
```diff
@@ -55,7 +74,13 @@ function showInfoPanel(d) {
         break;
       case "sensor":
         jsonToTable(data.value, "value");
-        break;
+            break;
```
Comment (andrew-buckley) [inline app/scripts/infopanel.js]: Check on the formatting here

File: app/scripts/graph.js [hunk 9]
```diff
@@ -429,15 +431,15 @@ function updateGraphVisualizer(source, animated) {
         .attr("width", nodeWidth)
         .attr("text-anchor", "start")
         .text(function (d) {
-            return d.label;
+            return d.label || d.name;
```
Comment (andrew-buckley) [inline app/scripts/graph.js:434]: The name *could* be added to the `extendNodeProperties` switch cases as the lowest priority option for the label. Just something to consider, not a blocker.

**Review by andrew-buckley — CHANGES_REQUESTED** [review]
This looks pretty solid, Ayo. Just go through and have a look at Lyrana's and my comments. A lot of the requests for changes surround general code formatting so like Lyrana suggested, take a few moments to give all of your changes a second look and check for formatting mismatches. VS and VSCode can help with this, but I can also understand that sometimes they don't follow existing formatting extremely well.

-andrew

**Pull request merged**